File: lib/productLookup.ts
import localDb from '@/data/local-products.json';
import type { MaterialId } from './data';

export interface Product {
  barcode: string;
  name?: string;
  brand?: string;
  image?: string;
  source: 'local' | 'openfoodfacts' | 'upcitemdb';
  packagingRaw?: string;
  notes?: string;
  inferredMaterials: MaterialId[];
}

const TAG_TO_MATERIAL: Record<string, MaterialId> = {
  'en:plastic': 'pet',
  'en:pet-1-polyethylene-terephthalate': 'pet',
  'en:pet': 'pet',
  'en:hdpe-2-high-density-polyethylene': 'hdpe',
  'en:hdpe': 'hdpe',
  'en:pvc-3-polyvinyl-chloride': 'pvc',
  'en:pvc': 'pvc',
  'en:ldpe-4-low-density-polyethylene': 'ldpe',
  'en:ldpe': 'ldpe',
  'en:pp-5-polypropylene': 'pp',
  'en:pp': 'pp',
  'en:ps-6-polystyrene': 'ps',
  'en:ps': 'ps',
  'en:other-plastics-7': 'otros',
  'en:glass': 'vidrio',
  'en:glass-bottle': 'vidrio',
  'en:glass-jar': 'vidrio',
  'en:paper': 'papel',
  'en:cardboard': 'carton',
  'en:carton': 'carton',
  'en:tetra-pak': 'tetrapak',
  'en:brick': 'tetrapak',
  'en:aluminium': 'latas_aluminio',
  'en:aluminium-can': 'latas_aluminio',
  'en:can': 'latas_acero',
  'en:steel': 'latas_acero',
  'en:tin': 'latas_acero',
};

function inferFromText(raw: string): MaterialId[] {
  const r = raw.toLowerCase();
  const found = new Set<MaterialId>();
  if (r.includes('vidrio') || r.includes('glass') || r.includes('botella vidrio')) found.add('vidrio');
  if (r.includes('cartón') || r.includes('carton') || r.includes('cardboard') || r.includes('caja')) found.add('carton');
  if (r.includes('papel') || r.includes('paper')) found.add('papel');
  if (r.includes('tetra') || r.includes('brick')) found.add('tetrapak');
  if (r.includes('aluminio') || r.includes('aluminium') || r.includes('lata de bebida') || r.includes('beverage can')) found.add('latas_aluminio');
  if (r.includes('lata') || r.includes('can') || r.includes('tin')) found.add('latas_acero');
  if (r.includes('plástico') || r.includes('plastic') || r.includes('pet') || r.includes('botella')) found.add('pet');
  if (r.includes('hdpe') || r.includes('pead') || r.includes('polietileno alta')) found.add('hdpe');
  if (r.includes('ldpe') || r.includes('pebd') || r.includes('bolsa')) found.add('ldpe');
  if (r.includes('polipropileno') || r.includes('polypropylene')) found.add('pp');
  return Array.from(found);
}

function tryLocal(barcode: string): Product | null {
  const productos = (localDb as any).productos as Record<
    string,
    { name: string; brand?: string; materials: MaterialId[]; notes?: string }
  >;
  const hit = productos[barcode];
  if (!hit) return null;
  return {
    barcode,
    name: hit.name,
    brand: hit.brand,
    source: 'local',
    notes: hit.notes,
    inferredMaterials: hit.materials,
  };
}

async function tryOpenFoodFacts(barcode: string): Promise<Product | null> {
  const url = `https://world.openfoodfacts.org/api/v2/product/${encodeURIComponent(
    barcode,
  )}.json?fields=product_name,brands,image_front_small_url,packaging,packaging_tags,categories`;
  const res = await fetch(url, { headers: { 'User-Agent': 'RecicLas/0.1 (Chile)' } });
  if (!res.ok) return null;
  const data = await res.json();
  if (data.status !== 1 || !data.product) return null;
  const p = data.product;
  const tags: string[] = p.packaging_tags || [];
  const found = new Set<MaterialId>();
  for (const tag of tags) {
    const m = TAG_TO_MATERIAL[tag];
    if (m) found.add(m);
  }
  inferFromText(`${p.packaging || ''} ${p.categories || ''}`).forEach((m) => found.add(m));
  return {
    barcode,
    name: p.product_name,
    brand: p.brands,
    image: p.image_front_small_url,
    source: 'openfoodfacts',
    packagingRaw: p.packaging,
    inferredMaterials: Array.from(found),
  };
}

async function tryUpcItemDb(barcode: string): Promise<Product | null> {
  const url = `https://api.upcitemdb.com/prod/trial/lookup?upc=${encodeURIComponent(barcode)}`;
  const res = await fetch(url);
  if (!res.ok) return null;
  const data = await res.json();
  const item = data?.items?.[0];
  if (!item) return null;
  const corpus = `${item.title || ''} ${item.description || ''} ${item.category || ''} ${item.brand || ''}`;
  return {
    barcode,
    name: item.title,
    brand: item.brand,
    image: item.images?.[0],
    source: 'upcitemdb',
    packagingRaw: item.description,
    inferredMaterials: inferFromText(corpus),
  };
}

export async function lookupProduct(barcode: string): Promise<Product | null> {
  const local = tryLocal(barcode);
  if (local) return local;
  try {
    const off = await tryOpenFoodFacts(barcode);
    if (off) return off;
  } catch {}
  try {
    const upc = await tryUpcItemDb(barcode);
    if (upc) return upc;
  } catch {}
  return null;
}

File: ios-native/RecicLas/Sources/Services/ProductLookup.swift
import Foundation

enum ProductLookup {

    /// Busca un producto por código de barra:
    /// 1) Catálogo local (productos chilenos curados)
    /// 2) Open Food Facts (mundial, gratis)
    /// 3) UPCitemdb trial (sin auth, ~100 req/día)
    static func lookup(barcode: String) async -> ProductLookupResult? {
        if let local = tryLocal(barcode: barcode) { return local }
        if let off = try? await tryOpenFoodFacts(barcode: barcode), off != nil { return off }
        if let upc = try? await tryUpcItemDb(barcode: barcode), upc != nil { return upc }
        return nil
    }

    private static func tryLocal(barcode: String) -> ProductLookupResult? {
        guard let p = Catalog.shared.localProducts[barcode] else { return nil }
        return ProductLookupResult(
            barcode: barcode,
            name: p.name,
            brand: p.brand,
            imageURL: nil,
            inferredMaterials: p.materials,
            notes: p.notes,
            source: .local
        )
    }

    private static func tryOpenFoodFacts(barcode: String) async throws -> ProductLookupResult? {
        let urlStr = "https://world.openfoodfacts.org/api/v2/product/\(barcode).json?fields=product_name,brands,image_front_small_url,packaging,packaging_tags,categories"
        guard let url = URL(string: urlStr) else { return nil }
        var req = URLRequest(url: url)
        req.setValue("RecicLas-iOS/0.1 (Chile)", forHTTPHeaderField: "User-Agent")
        let (data, _) = try await URLSession.shared.data(for: req)
        guard
            let raw = try JSONSerialization.jsonObject(with: data) as? [String: Any],
            let status = raw["status"] as? Int, status == 1,
            let product = raw["product"] as? [String: Any]
        else { return nil }

        let tags = (product["packaging_tags"] as? [String]) ?? []
        let packaging = (product["packaging"] as? String) ?? ""
        let categories = (product["categories"] as? String) ?? ""
        var found = Set<String>()
        for tag in tags {
            if let m = tagToMaterial[tag] { found.insert(m) }
        }
        for m in inferFromText("\(packaging) \(categories)") { found.insert(m) }

        return ProductLookupResult(
            barcode: barcode,
            name: product["product_name"] as? String,
            brand: product["brands"] as? String,
            imageURL: (product["image_front_small_url"] as? String).flatMap(URL.init),
            inferredMaterials: Array(found),
            notes: nil,
            source: .openFoodFacts
        )
    }

    private static func tryUpcItemDb(barcode: String) async throws -> ProductLookupResult? {
        let urlStr = "https://api.upcitemdb.com/prod/trial/lookup?upc=\(barcode)"
        guard let url = URL(string: urlStr) else { return nil }
        let (data, _) = try await URLSession.shared.data(from: url)
        guard
            let raw = try JSONSerialization.jsonObject(with: data) as? [String: Any],
            let items = raw["items"] as? [[String: Any]],
            let item = items.first
        else { return nil }

        let title = item["title"] as? String ?? ""
        let desc = item["description"] as? String ?? ""
        let category = item["category"] as? String ?? ""
        let brand = item["brand"] as? String ?? ""
        let corpus = "\(title) \(desc) \(category) \(brand)"

        let images = item["images"] as? [String] ?? []
        return ProductLookupResult(
            barcode: barcode,
            name: title,
            brand: brand.isEmpty ? nil : brand,
            imageURL: images.first.flatMap(URL.init),
            inferredMaterials: inferFromText(corpus),
            notes: nil,
            source: .upcItemDb
        )
    }

    private static let tagToMaterial: [String: String] = [
        "en:plastic": "pet",
        "en:pet-1-polyethylene-terephthalate": "pet",
        "en:pet": "pet",
        "en:hdpe-2-high-density-polyethylene": "hdpe",
        "en:hdpe": "hdpe",
        "en:pvc-3-polyvinyl-chloride": "pvc",
        "en:pvc": "pvc",
        "en:ldpe-4-low-density-polyethylene": "ldpe",
        "en:ldpe": "ldpe",
        "en:pp-5-polypropylene": "pp",
        "en:pp": "pp",
        "en:ps-6-polystyrene": "ps",
        "en:other-plastics-7": "otros",
        "en:glass": "vidrio",
        "en:glass-bottle": "vidrio",
        "en:paper": "papel",
        "en:cardboard": "carton",
        "en:tetra-pak": "tetrapak",
        "en:brick": "tetrapak",
        "en:aluminium": "latas_aluminio",
        "en:aluminium-can": "latas_aluminio",
        "en:can": "latas_acero",
        "en:steel": "latas_acero",
        "en:tin": "latas_acero"
    ]

    private static func inferFromText(_ raw: String) -> [String] {
        let r = raw.lowercased()
        var found = Set<String>()
        if r.contains("vidrio") || r.contains("glass") { found.insert("vidrio") }
        if r.contains("cartón") || r.contains("carton") || r.contains("cardboard") || r.contains("caja") { found.insert("carton") }
        if r.contains("papel") || r.contains("paper") { found.insert("papel") }
        if r.contains("tetra") || r.contains("brick") { found.insert("tetrapak") }
        if r.contains("aluminio") || r.contains("aluminium") || r.contains("beverage can") { found.insert("latas_aluminio") }
        if r.contains("lata") || r.contains("tin") { found.insert("latas_acero") }
        if r.contains("plástico") || r.contains("plastic") || r.contains("pet") || r.contains("botella") { found.insert("pet") }
        if r.contains("hdpe") || r.contains("pead") { found.insert("hdpe") }
        if r.contains("ldpe") || r.contains("pebd") || r.contains("bolsa") { found.insert("ldpe") }
        if r.contains("polipropileno") || r.contains("polypropylene") { found.insert("pp") }
        return Array(found)
    }
}
